### ISSUE TITLE
Workflow D.4 missing results

### ISSUE THREAD

**Issue by cbizon** [ranking-agent/strider#264]
tl;dr, this is a 4 hop query with both ends constrained.  If I run as one query, I get no results.  If I run as 2 2-hops, I find one node that should bridge the sides.

The 4 hop query is D.4_tryptophan-kynurenine-with-curie.json in the minihackathons.  It looks like:
```
{
    "message": {
        "query_graph": {
            "edges": {
                "e00": {
                    "subject": "n00",
                    "object": "n01"
                },
                "e01": {
                    "subject": "n01",
                    "object": "n02"
                },
                "e02": {
                    "subject": "n02",
                    "object": "n03"
                },
                "e03": {
                    "subject": "n03",
                    "object": "n04"
                }
            },
            "nodes": {
                "n00": {
                    "ids": [
                        "DRUGBANK:DB00150"
                    ],
                    "categories": [
                        "biolink:SmallMolecule"
                    ]
                },
                "n01": {
                    "categories": [
                        "biolink:Protein"
                    ]
                },
                "n02": {
                    "categories": [
                        "biolink:MolecularActivity"
                    ]
                },
                "n03": {
                    "categories": [
                        "biolink:MolecularActivity"
                    ]
                },
                "n04": {
                    "ids": [
                        "KEGG.COMPOUND:C02700"
                    ],
                    "categories": [
                        "biolink:SmallMolecule"
                    ]
                }
            }
        }
    },
    "log_level": "INFO"
}
```

Running this in prod strider:
```
strider returned in 0:02:36.727059s
Num Results: 0
Biothings Explorer ReasonerStdAPI took >60 seconds to respond (46 times)
Automat Covidkop KG (trapi v-1.1.2) took >60 seconds to respond (28 times)
Automat Robokop KG (trapi v-1.1.2) took >60 seconds to respond (49 times)
CAM-KP API took >60 seconds to respond (47 times)
Automat Textmining KP (trapi v-1.1.2) took >60 seconds to respond (20 times)
Automat Ontological Hierarchy (trapi v-1.1.2) took >60 seconds to respond (11 times)
Automat Cord19 (trapi v-1.1.2) took >60 seconds to respond (1 times)
Ontology-KP API took >60 seconds to respond (13 times)
RTX KG2 took >60 seconds to respond (18 times)
```

OTOH, if I first run the n0-n1-n2 side:
```
strider returned in 0:01:21.151440s
Num Results: 2493
Biothings Explorer ReasonerStdAPI took >60 seconds to respond (2 times)
CAM-KP API took >60 seconds to respond (1 times)
Automat Robokop KG (trapi v-1.1.2) took >60 seconds to respond (2 times)
```

And then I run the n4-n3-n2 side:
```
strider returned in 0:01:08.931279s
Num Results: 4
Ontology-KP API took >60 seconds to respond (1 times)
```

Comparing the n2 in each case, there is one that they have in common, so the original query should have returned an answer.

The only obvious difference is that there are some timeouts in the 4hop that are not in the others, specifically the KGX2 timeout.  I wonder if the timeout were relaxed a bit whether we would get results, or if this is a deeper problem

**Comment by patrickkwang** [comment]
This makes sense. By default, Strider will traverse this in the order in which the edges are provided: [n00, n01, n02, n03]. Around the third hop, the batches are getting really big, and KPs start timing out a lot. Increasing the timeout _might_ let you get through it, but I think a better solution is to implement the cleverer query planning we have discussed, which should traverse these in an order more like [n00, n03, n01, n02], regardless of the order in which they are provided. I'm working on that. But I'm curious what happens if you provide the four-hop right now and just shuffle the edges around to the way we want them.

**Comment by cbizon** [comment]
I guess it's a bit tricky to try that since edges is a {} not a [], so I don't know that order is going to be preserved all the way from my dict to yours.

**Comment by cbizon** [comment]
Wow, tried it out:
```
strider returned in 0:01:24.143231s
Num Results: 48
Biothings Explorer ReasonerStdAPI took >60 seconds to respond (1 times)
Ontology-KP API took >60 seconds to respond (4 times)
Automat Robokop KG (trapi v-1.1.2) took >60 seconds to respond (4 times)
```

Query planning FTW

**Comment by richakanwar13** [comment]
Can be closed as of 12/3.
